[PGNCCL] Add an API to get the status/error code of each PG

shuqiangzhang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #140087

Summary:
If unhealthy, the user should be able to get the type of errors, e.g.,
timeout,nccl error or remote error.

This API is applied to PG level, compared to the work.get_future_result() API which is applied to Work Level.
Error detection at PG level is much more convenient for users to handle the PG failure as a whole, e.g, restarting the PG.

Error handling at the work level is still useful for users to attach work specific context and debug the RC of the specific failing work/collective

Note it is critical for all ranks in the PG to be notified about an error as soon as it occurs, so we introduce an errorType of REMOTE_ERROR, which is 'broadcasted' from a src rank (which detects a local error) to all other ranks in the PG, the broadcast is done through TCPStore currently

Tags:

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

shuqiangzhang committed:
[PGNCCL] Add an API to check the healthiness of each PG
Summary:
If unhealthy, the user should be able to get the type of errors, e.g.,
timeout,nccl error or remote error.
Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

shuqiangzhang committed:
Update on "[PGNCCL] Add an API to check the healthiness of each PG"


Summary:
If unhealthy, the user should be able to get the type of errors, e.g.,
timeout,nccl error or remote error.
Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

shuqiangzhang committed:
Update on "[PGNCCL] Add an API to get the status/error code of each PG"


Summary:
If unhealthy, the user should be able to get the type of errors, e.g.,
timeout,nccl error or remote error.

This API is applied to PG level, compared to the work.get_future_result() API which is applied to Work Level.
Error detection at PG level is much more convenient for users to handle the PG failure as a whole, e.g, restarting the PG.

Error handling at the work level is still useful for users to attach work specific context and debug the RC of the specific failing work/collective

Tags:

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

shuqiangzhang committed:
Update on "[PGNCCL] Add an API to get the status/error code of each PG"


Summary:
If unhealthy, the user should be able to get the type of errors, e.g.,
timeout,nccl error or remote error.

This API is applied to PG level, compared to the work.get_future_result() API which is applied to Work Level.
Error detection at PG level is much more convenient for users to handle the PG failure as a whole, e.g, restarting the PG.

Error handling at the work level is still useful for users to attach work specific context and debug the RC of the specific failing work/collective

Tags:

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

shuqiangzhang committed:
Update on "[PGNCCL] Add an API to get the status/error code of each PG"


Summary:
If unhealthy, the user should be able to get the type of errors, e.g.,
timeout,nccl error or remote error.

This API is applied to PG level, compared to the work.get_future_result() API which is applied to Work Level.
Error detection at PG level is much more convenient for users to handle the PG failure as a whole, e.g, restarting the PG.

Error handling at the work level is still useful for users to attach work specific context and debug the RC of the specific failing work/collective

Note it is critical for all ranks in the PG to be notified about an error as soon as it occurs, so we introduce an errorType of REMOTE_ERROR, which is 'broadcasted' from a src rank (which detects a local error) to all other ranks in the PG, the broadcast is done through TCPStore currently

Tags:

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

kwen2501 commented on torch/_C/_distributed_c10d.pyi:
```diff
@@ -546,6 +546,12 @@ class _ProcessGroupWrapper(Backend):
+class ErrorType(Enum):
+    NO_ERROR = ...
+    TIMEOUT = ...
+    NCCL_ERROR = ...
```
Generalize `NCCL_ERROR` to `BACKEND_ERROR`?

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -146,6 +148,15 @@ enum ErrorHandlingMode {
   SkipCleanUp = 3
 };
 
+enum TORCH_API ErrorType {
+  NO_ERROR = 0,
+  TIMEOUT = 1,
+  NCCL_ERROR = 2,
+  // TODO, do we need to distinguish between remote timeout or remote NCCL
+  // errors?
+  REMOTE_ERROR = 3,
+};
+
```
I think we can define this API one level up in `Backend.hpp`.

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -757,6 +768,8 @@ class TORCH_API ProcessGroupNCCL : public Backend {
   // If all comms on this PG are fully initialized, return true.
   bool isInitialized();
 
+  ErrorType getError();
+
```
Same suggestion for uplifting. For non-NCCL backends, we can leave it as unimplemented.

shuqiangzhang replied:
Was debating about this with myself too, can make it in any backend

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -2058,6 +2112,11 @@ void ProcessGroupNCCL::watchdogHandler() {
       if (!terminateProcessGroup_.load()) {
         work.checkAndSetException();
       }
+      if (work.exception() && getError() == ErrorType::NO_ERROR) {
+        // set the error to the first error found
+        std::lock_guard<std::mutex> lock(errorMutex_);
+        error_ = ErrorType::NCCL_ERROR;
+      }
```
nit: lock -> read -> write to guarantee atomicity. 
Or, we can use std::atomic.

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -127,6 +127,8 @@ constexpr const char* NCCL_BACKEND_NAME = "nccl";
 
 constexpr const char* EXCEPTION_DUMP = "exception_dump";
 
+constexpr const char* REMOTE_ERROR_SIGNAL = "remote_error";
```
nit: `storeDumpKey`, `storeErrorSignalKey`

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -2046,6 +2086,20 @@ void ProcessGroupNCCL::watchdogHandler() {
 
+    // Check and set remote error if it has not been set before
+    if (getError() == ErrorType::NO_ERROR) {
+      int remoteErrorRank =
+          getSignalSrcRank(store_, std::string(REMOTE_ERROR_SIGNAL));
```
I haven't checked whether `getSignalSrcRank` could be blocking or not. Let's be careful when putting potentially blocking call in watchdog.

shuqiangzhang replied:
It's non blocking in a sense, we first 'check' if key exists which is nonblocking, and then read/get the value only if the key exists. Otherwise, if we try to get the kv directly, it could be blocking.

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -2079,6 +2142,9 @@ void ProcessGroupNCCL::watchdogHandler() {
             ", last completed work: ",
             pgStatus_->lastCompletedSeq);
 
+        // broadcast remote error signal to all other ranks in this specific PG.
+        broadcastSignal(store_, std::string(REMOTE_ERROR_SIGNAL), rank_);
+
```
I haven't checked whether `broadcastSignal` could be blocking or not. Let's be careful when putting potentially blocking call in watchdog.

shuqiangzhang replied:
The same functionality was in watchdog thread in the original code

kwen2501 reviewed (commented):
Overall looks good to me.

shuqiangzhang commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1948,38 +1953,73 @@ bool ProcessGroupNCCL::verifyWorkTimeoutForTest(
 
-// Broadcast flight-recorder dump signal
-void ProcessGroupNCCL::broadcastDumpSignal() {
+void ProcessGroupNCCL::broadcastSignal(
+    c10::intrusive_ptr<Store>& store,
+    const std::string& signal,
+    int srcRank) {
```
will keep it in PGNCCL first, we can move it later if other backends need it

shuqiangzhang committed:
Update on "[PGNCCL] Add an API to get the status/error code of each PG"


Summary:
If unhealthy, the user should be able to get the type of errors, e.g.,
timeout,nccl error or remote error.

This API is applied to PG level, compared to the work.get_future_result() API which is applied to Work Level.
Error detection at PG level is much more convenient for users to handle the PG failure as a whole, e.g, restarting the PG.

Error handling at the work level is still useful for users to attach work specific context and debug the RC of the specific failing work/collective

Note it is critical for all ranks in the PG to be notified about an error as soon as it occurs, so we introduce an errorType of REMOTE_ERROR, which is 'broadcasted' from a src rank (which detects a local error) to all other ranks in the PG, the broadcast is done through TCPStore currently

Tags:

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

shuqiangzhang committed:
Update on "[PGNCCL] Add an API to get the status/error code of each PG"


Summary:
If unhealthy, the user should be able to get the type of errors, e.g.,
timeout,nccl error or remote error.

This API is applied to PG level, compared to the work.get_future_result() API which is applied to Work Level.
Error detection at PG level is much more convenient for users to handle the PG failure as a whole, e.g, restarting the PG.

Error handling at the work level is still useful for users to attach work specific context and debug the RC of the specific failing work/collective

Note it is critical for all ranks in the PG to be notified about an error as soon as it occurs, so we introduce an errorType of REMOTE_ERROR, which is 'broadcasted' from a src rank (which detects a local error) to all other ranks in the PG, the broadcast is done through TCPStore currently

Tags:

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

shuqiangzhang committed:
Update on "[PGNCCL] Add an API to get the status/error code of each PG"


Summary:
If unhealthy, the user should be able to get the type of errors, e.g.,
timeout,nccl error or remote error.

This API is applied to PG level, compared to the work.get_future_result() API which is applied to Work Level.
Error detection at PG level is much more convenient for users to handle the PG failure as a whole, e.g, restarting the PG.

Error handling at the work level is still useful for users to attach work specific context and debug the RC of the specific failing work/collective

Note it is critical for all ranks in the PG to be notified about an error as soon as it occurs, so we introduce an errorType of REMOTE_ERROR, which is 'broadcasted' from a src rank (which detects a local error) to all other ranks in the PG, the broadcast is done through TCPStore currently

Tags:

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

shuqiangzhang committed:
Update on "[PGNCCL] Add an API to get the status/error code of each PG"


Summary:
If unhealthy, the user should be able to get the type of errors, e.g.,
timeout,nccl error or remote error.

This API is applied to PG level, compared to the work.get_future_result() API which is applied to Work Level.
Error detection at PG level is much more convenient for users to handle the PG failure as a whole, e.g, restarting the PG.

Error handling at the work level is still useful for users to attach work specific context and debug the RC of the specific failing work/collective

Note it is critical for all ranks in the PG to be notified about an error as soon as it occurs, so we introduce an errorType of REMOTE_ERROR, which is 'broadcasted' from a src rank (which detects a local error) to all other ranks in the PG, the broadcast is done through TCPStore currently

Tags:

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

shuqiangzhang committed:
Update on "[PGNCCL] Add an API to get the status/error code of each PG"


Summary:
If unhealthy, the user should be able to get the type of errors, e.g.,
timeout,nccl error or remote error.

This API is applied to PG level, compared to the work.get_future_result() API which is applied to Work Level.
Error detection at PG level is much more convenient for users to handle the PG failure as a whole, e.g, restarting the PG.

Error handling at the work level is still useful for users to attach work specific context and debug the RC of the specific failing work/collective

Note it is critical for all ranks in the PG to be notified about an error as soon as it occurs, so we introduce an errorType of REMOTE_ERROR, which is 'broadcasted' from a src rank (which detects a local error) to all other ranks in the PG, the broadcast is done through TCPStore currently

Tags:

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

kwen2501 reviewed (approved):
LGTM! Great achievement towards user initiated abort/restart.

shuqiangzhang commented:
@pytorchbot merge

shuqiangzhang commented:
@pytorchbot merge -f "mege timeout"

shuqiangzhang commented:
This PR has caused perf drop in some tests, this is most likely resulted from TCPStore read in watchdog thread. Moving it to another thread, e.g., monitor thread which has been already doing tcpstore reads, could avoid the perf drop. Will do it in another fresh PR, since this PR has been old and conflicts resolving is needed
